### PR TITLE
Feat/auth next state machine queue

### PR DESCRIPTION
**Pull request by haverchuck**
#### Description of changes
This PR adds the following:
1. An event queue on the machine class. This queue should have the behavior of enqueing events and ceding control back to the call site, and then dequeing subsequent events once the first is fully processed. It's using setInterval to do this, which seems to work, but it's been a while since I've done something like this so I am open to better ideas!
2. Hub listeners to subscribe to state machines
3. Other various changes, such as allowing Promises as an invocation type.

It's a litttle messy still, but wanted to get the PR up.

#### Description of how you validated changes
Unit tests.

#### Checklist

- [X] PR description included
- [ ] `yarn test` passes
- [X] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/stateMachine/machine.ts [hunk 1]
```diff
@@ -1,68 +1,140 @@
...
+	}
+
+	private async _queueProcessor() {
+		this._interval = setInterval(async () => {
+			if (this._queue.length > 0 && !this._eventInFlight) {
+				let currentEvent = this._queue.shift();
+				this._eventInFlight = true;
+				await this._handleQueuedItem(currentEvent!);
+			}
+		}, 1);
```
Comment (AllanZhengYP) [inline packages/auth/src/stateMachine/machine.ts:69]: I have seen the use of `setTimeout(callback, 0)` more often to 'nextTick' the event loop. The `0` timeout means immediately. See the reference: https://developer.mozilla.org/en-US/docs/Web/API/setTimeout#delay

Nice thing of it is that you don't need to call `clearTimeout()` after the callback is fired.

File: packages/auth/src/stateMachine/types.ts [hunk 2]
```diff
@@ -1,28 +1,52 @@
...
 
 /**
- * Base type for a MachineEvent's payload
+ * The type accepted by Machine's send method
  * @typeParam PayloadType - The type of the Event's payload
+ * @typeParam ContextType - The type of the target Machine's context
  * @param name - The event name; used when matching a transition
  * @param payload - The event payload
+ * @param restingStates - The state names of the underlying machine which, when reached, will allow for the event to be dequeued
+ * @param completer - A promise that will resolve when a restingState is reached
```
Comment (AllanZhengYP) [inline packages/auth/src/stateMachine/types.ts:25]: We need to think more on whether resting states are necessary. This property conflicts with the state machine completion, which happens when current state effects are completed and no more state transitions can be made.

File: packages/auth/src/stateMachine/machine.ts [hunk 3]
```diff
@@ -1,95 +1,207 @@
...
+		});
+	}
 
-		///TODO: Communicate null transition
-		if (!validTransition) return;
-		const checkGuards = this._checkGuards(validTransition, event);
-		//TODO: Communicate guard failure
-		if (!checkGuards) return;
+	/**
+	 * Returns the current state after previously queued events have been flushed
```
Comment (AllanZhengYP) [inline packages/auth/src/stateMachine/machine.ts:69]: nit: `flushed` may not be the most precise word. Normally flush means getting rid of the previous events. I don't think it's the case here. Maybe this is better:
```suggestion
	 * Returns the current state after the state machine has completed reacting to previously queued events
```

File: packages/auth/src/stateMachine/machine.ts [hunk 4]
```diff
@@ -1,95 +1,207 @@
...
+		const validTransition = this._current.findTransition(event);
+		if (!validTransition) {
+			this._handleFailure(StateMachineHubEventName.NULL_TRANSITION, event);
+			return;
+		}
+		const checkGuards = this._checkGuards(validTransition, event);
+		if (!checkGuards) {
+			this._handleFailure(StateMachineHubEventName.STATE_GUARD_FAILURE, event);
+			return;
+		}
```
Comment (AllanZhengYP) [inline packages/auth/src/stateMachine/machine.ts:142]: Why are guards required?

Comment (haverchuck) [reply packages/auth/src/stateMachine/machine.ts:142]: They make state transitions conditional. For example, we could use a guard to fail an attempt to  use Hosted UI if the config doesn't contain the appropriate keys

**Pull request closed without merging**